Ex2

lahavrom opened the pull request:
refactored code to use classes and methods.
when user inputs number it adds a task to catch the pokemon whose id is the number.
also works when input is a list of numbers separated by comma.
handles errors and "bad" user inputs.
also fetches the pokemon's type.
still have the functionalities to sort tasks by name/date and to clear all.

@amirbardugo @eranbenmoha

lahavrom committed:
feat: when input number it fetches from pokeapi

lahavrom committed:
fix: handles when input ends or start with comma

lahavrom committed:
fix: works locally without imports/exports

amirbardugo commented on src/ItemManager.js:
```diff
@@ -0,0 +1,52 @@
+
```
for the next one I think it will be easier for you if you'll save each exercise in its own folder, and just duplicate it when moving from one to the other, this way you could compare them and go back to previous steps more easily

amirbardugo commented on src/PokemonClient.js:
```diff
@@ -0,0 +1,41 @@
+        } catch (err) {
+            // console.log(err);
+            return ["Pokemon with ID " + id + " was not found", false];
+        }
```
in terms of responsibilities, does it really make sense that PokemonClient will be responsible for the exact message the client will see, if we'll use this function somewhere else and want different message
I think in this case throwing a proper exception makes more sense, this way anyone can catch it and do whatever he wants.
Also, this way you won't need that second argument in the return value and the API will be clearer

amirbardugo commented on src/PokemonClient.js:
```diff
@@ -0,0 +1,41 @@
+
+    async getListPokemon(id_array) {
+        try {
+            const response = await Promise.all(id_array.map(elem => {
```
use better naming, when i saw elem i thought you mean some Dom element, 
something like pokemonID will make it much clearer

amirbardugo commented on src/PokemonClient.js:
```diff
@@ -0,0 +1,41 @@
+        } catch(err) {
+            // console.log(err);
+            return ["Failed to fetch pokemons with this input: " + id_array, false];
+        }
```
the same comment as before, throwing an exception in case of failures is a clearer API

amirbardugo commented on src/Main.js:
```diff
@@ -0,0 +1,177 @@
+            }
+        });
+        const clearAllBtn = document.getElementById("clearAll");
+        clearAllBtn.addEventListener("click", () => {
+            this.clearAll();
+        });
```
just a shorter version for those cases 😃 
```suggestion
        clearAllBtn.addEventListener("click", this.clearAll);
```

amirbardugo commented on src/ItemManager.js:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    async handleTask(task) {
+        let taskSplit = task.split(",");
```
In terms of responsibilities this logic should be on your main.js, the item manager should not be familiar with the UI, the structure of the input or anything, I would want that if you will drastically change the UI item manager would still be relevant and won't be affected

amirbardugo commented on src/ItemManager.js:
```diff
@@ -0,0 +1,52 @@
+    async handleTask(task) {
+        let taskSplit = task.split(",");
+        if (isNaN(taskSplit[0])) {
+            return [[task], false];
```
very confusing to read such a return value, I have no idea what false stands for, I need to search for the usages to understand that

amirbardugo commented on src/ItemManager.js:
```diff
@@ -0,0 +1,52 @@
+            return [[taskName], true];
+        }
+        // handle list of pokemons
+        const result = await this.pokemonClient.getListPokemon(taskSplit);
+        if (!result[1]) {
+            return [[result[0]], true]
+        }
```
also would be nicer with throwing an exception you could just use try catch here instead of this

amirbardugo commented on src/ItemManager.js:
```diff
@@ -0,0 +1,52 @@
+        let pokemons = [];
+        for (let p of result[0]) {
+            const pokemon = await p;
+            const taskName = `Catch ${await pokemon.name}, the ${pokemon.types[0].type.name} type pokemon`;
```
this line is duplicated i would extract that to a private function here in the class

amirbardugo commented on src/ItemManager.js:
```diff
@@ -0,0 +1,52 @@
+
+    // add to array
+    addTask(task, taskId, pokemon) {
+        if (this.tasks.filter(e => (pokemon && e.task === task)).length > 0){
```
you can use find or find index instead of filter, filter will go over all the array instead of stopping on the first result

amirbardugo commented on src/ItemManager.js:
```diff
@@ -0,0 +1,52 @@
+    // add to array
+    addTask(task, taskId, pokemon) {
+        if (this.tasks.filter(e => (pokemon && e.task === task)).length > 0){
+            return false;
```
also here, its better to use exceptions, i think its more clear and more expected from such classes

amirbardugo commented on src/Main.js:
```diff
@@ -0,0 +1,177 @@
+
+class Main {
```
you didn't implement a separate render method, that would simplify this file a lot.
this way you can have only one or two dedicated functions that manipulates the dom and renders the whole list, so after reorder, removing one, removing all or adding 4 new tasks you would call the same function - renderList that can handle it all by simply rerender the whole list (removing and building it again)

lahavrom replied:
I understand, but it isn't inefficient? each adding/removing/reordering to remove and render **all** of the tasks?

amirbardugo reviewed (commented):
Great exercise overall, check out my comments I think some of them can help you write better and much clearer code, as always feel free to ping me for any questions 😃

lahavrom committed:
feat: Main responsible for all UI.
throw catch mechanism when handling errors.
render tasks function based on itemManager
